### PR TITLE
Cache cleanup and bug fixes

### PR DESCRIPTION
**Pull request by oldalton**
This addresses following issues:
1. Remove expired access tokens
2. Remove invalid refresh tokens
3. Only return accounts with tokens in cache for that client or family id
4. Disassociate application from its family ID after account removal, so that app cannot continue using family refresh token without interactive login first.
5. Removed unnecessary cache accessor methods/minor code cleanup

MSAL PR: https://github.com/AzureAD/microsoft-authentication-library-for-objc/pull/420

### CODE REVIEW

File: IdentityCore/tests/integration/ios/MSIDDefaultSilentTokenRequestTests.m [hunk 1]
```diff
@@ -340,6 +341,115 @@ - (void)testAcquireTokenSilent_whenExpiredAccessTokenInCache_shouldRefreshToken
...
+    MSIDAccessToken *accessToken = [tokenCache getAccessTokenForAccount:accountIdentifier configuration:silentParameters.msidConfiguration context:nil error:nil];
+    XCTAssertNotNil(accessToken);
+}
+
+- (void)testAcquireTokenSilent_whenATExpired_AndFailedToRefreshToken_shouldReturnError_AndRemoveExpiredAccesstoken_AndKeepRefreshTokenInCache
+{
+    MSIDRequestParameters *silentParameters = [self silentRequestParameters];
+    MSIDDefaultTokenCacheAccessor *tokenCache = self.tokenCache;
+
+    [self saveExpiredTokensInCache:tokenCache configuration:silentParameters.msidConfiguration];
```
Comment (unpluggedk) [inline IdentityCore/tests/integration/ios/MSIDDefaultSilentTokenRequestTests.m]: nit nit: quick test method to test if this actually works? I think multiple times, we rely on this being working yet I don't think we test this.

File: IdentityCore/src/cache/MSIDCacheAccessor.h [hunk 2]
```diff
@@ -96,8 +82,18 @@
                                              error:(NSError **)error;
 
 - (BOOL)clearCacheForAccount:(MSIDAccountIdentifier *)account
+                   authority:(MSIDAuthority *)authority
                     clientId:(NSString *)clientId
+                    familyId:(NSString *)familyId
                      context:(id<MSIDRequestContext>)context
                        error:(NSError **)error;
 
+- (BOOL)validateAndRemoveRefreshToken:(MSIDBaseToken<MSIDRefreshableToken> *)token
```
Comment (unpluggedk) [inline IdentityCore/src/cache/MSIDCacheAccessor.h]: nit: is it really "validating" anything?

**Comment by unpluggedk** [comment]
Looks mostly fine to me except few nits

File: IdentityCore/src/cache/accessor/MSIDDefaultTokenCacheAccessor.m [hunk 3]
```diff
@@ -279,32 +254,91 @@ - (MSIDIdToken *)getIDTokenForAccount:(MSIDAccountIdentifier *)account
...
+    accountsQuery.username = accountIdentifier.legacyAccountId;
+
+    NSArray<MSIDAccountCacheItem *> *allAccounts = [_accountCredentialCache getAccountsWithQuery:accountsQuery context:context error:error];
 
-    if (!filteredAccountsSet)
+    if (!allAccounts)
     {
         MSID_LOG_ERROR(context, @"(Default accessor) Failed accounts lookup");
         [MSIDTelemetry stopCacheEvent:event withItem:nil success:NO context:context];
         return @[];
```
[the remark this answers is not in the document]
Comment (oldalton) [reply IdentityCore/src/cache/accessor/MSIDDefaultTokenCacheAccessor.m]: Fixed :)

File: IdentityCore/src/cache/accessor/MSIDDefaultTokenCacheAccessor.m [hunk 4]
```diff
@@ -279,32 +254,91 @@ - (MSIDIdToken *)getIDTokenForAccount:(MSIDAccountIdentifier *)account
...
+        refreshTokenQuery.environmentAliases = environmentAliases;
+        refreshTokenQuery.clientIdMatchingOptions = MSIDSuperSet;
+
+        NSArray<MSIDCredentialCacheItem *> *refreshTokens = [_accountCredentialCache getCredentialsWithQuery:refreshTokenQuery context:context error:error];
+
+        if (!refreshTokens)
+        {
+            MSID_LOG_ERROR(context, @"(Default accessor) Failed refresh token lookup");
+            [MSIDTelemetry stopCacheEvent:event withItem:nil success:NO context:context];
+            return @[];
```
[the remark this answers is not in the document]
Comment (oldalton) [reply IdentityCore/src/cache/accessor/MSIDDefaultTokenCacheAccessor.m]: Fixed

File: IdentityCore/src/cache/accessor/MSIDDefaultTokenCacheAccessor.m [hunk 5]
```diff
@@ -279,32 +254,91 @@ - (MSIDIdToken *)getIDTokenForAccount:(MSIDAccountIdentifier *)account
...
+
+        filterAccountIds = [NSSet setWithArray:[refreshTokens valueForKey:@"homeAccountId"]];
+    }
+
+    NSMutableSet<MSIDAccount *> *filteredAccountsSet = [NSMutableSet new];
+
+    for (MSIDAccountCacheItem *accountCacheItem in allAccounts)
+    {
+        // If we have accountIds to filter by, only return account if it has an associated refresh token
+        if (!filterAccountIds || [filterAccountIds containsObject:accountCacheItem.homeAccountId])
```
[the remark this answers is not in the document]
Comment (oldalton) [reply IdentityCore/src/cache/accessor/MSIDDefaultTokenCacheAccessor.m]: Fixed

File: IdentityCore/tests/integration/ios/MSIDDefaultSilentTokenRequestTests.m (hunk 1 above)
Comment (oldalton) [reply IdentityCore/tests/integration/ios/MSIDDefaultSilentTokenRequestTests.m]: I checked implementation and it's using our accessor's method saveTokensWithConfiguration, which is covered with unit tests.

File: IdentityCore/src/cache/MSIDCacheAccessor.h (hunk 2 above)
Comment (oldalton) [reply IdentityCore/src/cache/MSIDCacheAccessor.h]: It validates that the refresh token hasn't been changed by other app :)

**Pull request merged**